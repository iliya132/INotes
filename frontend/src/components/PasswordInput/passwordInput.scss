@import '../../Misc/colors.scss';

.input-container {
    display: flex;
    width: 100%;
    height: 30px;
    border-bottom: 1px solid $TextColor;

    .input {
        margin-left: 5px;
        width: 100%;
        height: 30px;
        border: none;
        outline: none;
        background-color: transparent;
        color: $TextColor;
        font-family: 'Montserrat Medium';
        font-size: 14px;
        line-height: 13px;
    }
}

.pre-icon{
    margin: auto;
}

.eye-icon {
    display: inline-block;
    grid-column: 2;
    width: 48px;
    height: 48px;
    margin-top: -8px;
    cursor: pointer;
}

.error-label {
    max-height: 100%;
    color: $Error;
    font-size: 9px;
    font-weight: 400;
    overflow-wrap: break-word;
}

.input-error {
    color: $Error !important;
}
