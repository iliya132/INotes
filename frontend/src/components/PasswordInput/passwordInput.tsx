import React, { useState } from 'react';
import IInputProps from './types';
import styles from './passwordInput.scss';
import Svg from '../Svg';
import { Icons } from '../Svg/types';
import classNames from 'classnames';

export function PasswordInput(props: IInputProps) {
    const { id, error, autocomplete, className, onChange, placeholder, name } = props;
    const [isPasswordShown, setPasswordShown] = useState(false);
    const inputType = isPasswordShown ? 'text' : 'password';
    const internalClassName = error
        ? classNames(styles['input'], styles['input-error'], className)
        : classNames(styles['input'], className);
    return (
        <>
            <div className={styles['input-container']}>
                <Svg icon={Icons.Key} className={styles['pre-icon']} />
                <input
                    className={internalClassName}
                    id={id}
                    name={name}
                    type={inputType}
                    onChange={onChange}
                    autoComplete={autocomplete}
                    placeholder={placeholder}
                />
                {GetIcon(isPasswordShown, () => setPasswordShown(!isPasswordShown))}
            </div>
            <label className={styles['error-label']} htmlFor={id}>
                {error}
            </label>
        </>
    );
}

function GetIcon(showPassword: boolean, onclick: () => void) {
    return (
        <>
            {showPassword ? (
                <Svg icon={Icons.EyeClosed} className={styles['eye-icon']} onClick={onclick} />
            ) : (
                <Svg icon={Icons.EyeOpened} className={styles['eye-icon']} onClick={onclick} />
            )}
        </>
    );
}
