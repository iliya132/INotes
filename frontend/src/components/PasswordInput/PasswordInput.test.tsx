/* eslint-disable no-undef */
import React from 'react';
import { createRoot, Root } from 'react-dom/client';
import { act } from 'react-dom/test-utils';
import { PasswordInput } from './passwordInput';

//#region common setup
let root: Root | null = null;
let container: Element | null = null;

beforeEach(() => {
    act(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
        root = createRoot(container);
    });
});

afterEach(() => {
    act(() => {
        root!.unmount();
        container!.remove();
        container = null;
    });
});
//#endregion

describe('test password input component', () => {
    it('can render', () => {
        act(() => {
            root!.render(<PasswordInput id='test-id'/>)
        })
    });

    it('renders key icon', () => {
        act(() => {
            root!.render(<PasswordInput id='test-id'/>)
        })

        const keyIcon = container!.querySelector("svg")
        expect(keyIcon).not.toBeNull()
        expect(keyIcon!.innerHTML).toContain("#key")
    });

    it("matches snapshot", () => {
        act(() => {
            root!.render(<PasswordInput id='test-id'/>)
        })
        expect(container!.firstChild).toMatchSnapshot("Password input test snapshot")
    })

    it("can show and hide password", () => {
        act(() => {
            root!.render(<PasswordInput id='test-id'/>)
        })

        const eyeIcon = container!.querySelector(".eye-icon")
        const input = container!.querySelector("input");
        expect(input).not.toBeNull();
        expect(eyeIcon).not.toBeNull();

        expect(input!.type).toEqual("password")
        expect(eyeIcon?.innerHTML).toContain("#eye-open")

        act(() => {
            eyeIcon?.dispatchEvent(new MouseEvent("click", {bubbles: true}))
        })

        expect(input!.type).toEqual("text")
        expect(eyeIcon?.innerHTML).toContain("#eye-closed")
    })

    it("show error", () => {
        act(()=> {
            root!.render(<PasswordInput id="test-id" error='test-error'/>)
        })

        const expectedInputClassName = "input input-error"
        const errorLabel = container!.querySelector("label")
        expect(errorLabel?.textContent).toEqual("test-error")

        const input = container!.querySelector("input")
        expect(input?.className).toEqual(expectedInputClassName)
    })
});
