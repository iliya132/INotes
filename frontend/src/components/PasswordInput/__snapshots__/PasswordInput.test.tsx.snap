// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test password input component matches snapshot: Password input test snapshot 1`] = `
<div
  class="input-container"
>
  <svg
    class="image pre-icon tooltip"
    fill="BlackCommon"
    height="17"
    width="17"
  >
    <use
      xlink:href="test-file-stub#key"
    />
  </svg>
  <input
    class="input"
    id="test-id"
    type="password"
  />
  <svg
    class="image eye-icon tooltip"
    fill="BlackCommon"
    height="48"
    width="48"
  >
    <use
      xlink:href="test-file-stub#eye-open"
    />
  </svg>
</div>
`;
